Chapter-4-applying-tactical-domain-driven-design/README.adoc

```diff
@@ -40,29 +77,238 @@ IMPORTANT: It makes no sense to consider the Domain Model in typical CRUD module
+- SSN (Social Security Number) in the financial domain
+- Money in the financial domain (amount and currency)
+- Address in the e-commerce domain
+- PhoneNumber in the telecommunication domain
+- Email in the e-commerce domain
+- Energy Indicator in the eco domain (amount and unit)
+
+==== Main characteristics of value objects:
+- They have no identity
+- They represent a bussiness concept
+- They can encapsulate bussiness logic like validation during object initialization
+- They can be used as a part of an entity or aggregate root
+- They are immutable
+- They are compared by their properties
+- They have equals and hashcode methods implemented
+- In one bounded context concept can be modeled as value object and in another as entity
+
+[source,csharp]
+----
+public sealed class Signature : ValueObject
+{
+    public string Value { get; init; }
+    public DateTimeOffset SignedAt { get; init; }
+
+    private Signature(string value)
+    {
+        if (string.IsNullOrWhiteSpace(value))
+            throw new ArgumentException("Signature cannot be empty");
+
+        Value = value;
+    }
+
+    public static Signature Create(string value) => new(value);
+
+    protected override IEnumerable<object> GetEqualityComponents()
+    {
+        yield return Value;
+    }
+}
+----
+
+[[aggregate-root]]
+==== Aggregate Root
+
+An **Aggregate Root** is a collection of related objects that are treated as a single unit. It consists of entities and value objects that are interconnected through a root entity. **Consider the aggregate root as a safeguard for invariants, ensuring the consistency and validity of the domain model.** Those invariants are esential bussiness rules, we can't model real business processes without them.
+
+Aggregate roots are also transactional boundaries. All changes to the aggregate should be done through the aggregate root. This ensures that the aggregate is always in a consistent state.
+
+image::Assets/aggregate_root_internals.png[]
+
+NOTE: When you are looking for boundaries of an aggregate, you should consider which data has to change together (has to be in the same transaction) and which data can be changed independently.
+
+==== Main characteristics of aggregate roots:
+- They enforce business rules and invariants internally
+- They have a unique identity
+- They encapsulate entities and value objects and protect them from direct access
+- They are transactional boundaries
+- They have a lifecycle
+- They are lightweight
+- Each method execution state change is wrapped in a transaction
+- They can raise domain events after creating or state change after creating or state change
+
+[[domain-events]]
+==== Domain Events
+
+**Domain Events** are used to capture and communicate important events that occur within the domain. These events can trigger side effects or workflows in other parts of the system. They can also be translated into messages and sent to an external system.
```

Comment:
They don't trigger workflows in other parts of the system - this is a task of integration (external, public) events

Reply:
Thanks that is really bad.